compact_blank (#137)
andycroll/andycroll.com: _posts/2021/2021-09-20-compact_blank-remove-empty-strings-from-array-and-hash.md

---
title: "Use compact_blank to remove empty strings from Arrays and Hashes"
description: "This is one of those “Nice Things” in Rails"
layout: article
category: ruby
image:
  base: '2021/compact_blank-remove-empty-strings-from-array-and-hash'
  alt: "Compact Disc"
  credit: "Jan Huber"
  source: "https://unsplash.com/photos/oTCRizM-PUI"

---

> Active Support is a library of “useful” things that supports (!) the other parts of Rails. It includes ways to [extend the base Ruby classes](https://guides.rubyonrails.org/active_support_core_extensions.html), such as `Hash`, `String` and `Integer`.

Reviewer: *that support

Author: ```suggestion
Active Support is a library of “useful” things that support (!) the other parts of Rails. It includes ways to [extend the base Ruby classes](https://guides.rubyonrails.org/active_support_core_extensions.html), such as `Hash`, `String` and `Integer`.
```

> Even the most experienced Rails developer is constantly finding helpful methods in Active Support that are new to them. And useful methods are continually being added in new versions of the library.

Reviewer: *constantly discovering helpful methods... (don't need the 'that are new to them' part then, I think).

Author: ```suggestion
Even the most experienced Rails developer is constantly finding helpful methods in Active Support. And useful methods are continually being added in new versions of the library.
```

> We often have to clean up the values of`Hash`es and `Array`s when parsing paramters, or if our code accepts complex input. Ruby provides the `#compact` method to remove `nil` values from an array and (since 2.4) has provided a [the same `#compact` method](https://ruby-doc.org/core-2.4.0/Hash.html#method-i-compact) for Hash.

Reviewer: *parameters

*has provided the same

Author: ```suggestion
We often have to clean up the values of`Hash`es and `Array`s when parsing parameters, or if our code accepts complex input. Ruby provides the `#compact` method to remove `nil` values from an array and (since 2.4) has provided [the same `#compact` method](https://ruby-doc.org/core-2.4.0/Hash.html#method-i-compact) for Hash.
```

> In addition to `nil` often we’re looking to remove empty strings or other “blank” objects as well.

Reviewer: *we're often looking to remove...

Author: ```suggestion
In addition to `nil` often we’re often looking to remove empty strings or other “blank” objects as well.
```

#=> [1, “3”]

[1, nil, “3”, “”, []].reject(&:blank?)
#=> [1, “3”]

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.compact
#=> {1 => 2, “3” => “4”, 4 => “”, 5 => []}

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.reject { |_, v| v.nil? || v&.empty? }
#=> {1 => 2, “3” => “4”}

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.reject { |_, v| v.blank? }
#=> {1 => 2, “3” => “4”}
```

## Use…

…Active Support’s `#compact_blank` method:

```ruby
[1, nil, “3”, “”, []].compact_blank
#=> [1, “3”]

result = {1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.compact_blank
#=> {1 => 2, “3” => “4”}
```

The [#compact_blank](https://api.rubyonrails.org/classes/Enumerable.html#method-i-compact_blank) method was [added in Rails 6.1](https://github.com/rails/rails/blob/v6.1.4/activesupport/CHANGELOG.md#rails-610-december-09-2020).

## Why?

This is a “useful Rails thing” that feels like it’s part of Ruby once you’re used to it. Many of the core extensions in Active Support have that feeling in common. It reads a _lot_ better than the `#reject`-based versions.

## Why not?

Some folks resent the intrusion of the framework into the base concepts of the language.

> However, there have definitely been cases where Ruby has adopted an “Active Support-ism”. `Hash#compact` was in Rails long before it was in the base language.

Reviewer: This reads so much better now! 😁

---
title: "Use compact_blank to remove empty strings from Arrays and Hashes"
description: "This is one of those “Nice Things” in Rails"
layout: article
category: ruby
image:
  base: '2021/compact_blank-remove-empty-strings-from-array-and-hash'
  alt: "Compact Disc"
  credit: "Jan Huber"
  source: "https://unsplash.com/photos/oTCRizM-PUI"

---

> Active Support is a mixed library of “useful” things that’s included with Rails. This includes [extending the base Ruby classes](https://guides.rubyonrails.org/active_support_core_extensions.html) such as `Hash`, `String` and `Integer`.

Reviewer: *is a Rails library of "useful" things...?

* This includes ways to extend the base Ruby classes, such as...

Author: ```suggestion
Active Support is a library of “useful” things that supports (!) the other parts of Rails. It includes ways to [extend the base Ruby classes](https://guides.rubyonrails.org/active_support_core_extensions.html), such as `Hash`, `String` and `Integer`.
```

> Even the most experienced Rails developer is constantly finding helpful methods that are new to them. And there are useful new methods being added even in recent versions.

Reviewer: ```suggestion
Even the most experienced Rails developer is constantly finding helpful methods in Active Support that are new to them. And useful methods are continually being added in new versions of the library.
```

> A task I often find myself doing is cleaning up the values of`Hash`es and `Array`s. Ruby provides the `compact` method to remove `nil` values from an array and (since 2.4) has provided a [similarly named method](https://ruby-doc.org/core-2.4.0/Hash.html#method-i-compact) for Hash.

Reviewer: *similarly-named (I think)

Author: ```suggestion
A task I often find myself doing is cleaning up the values of`Hash`es and `Array`s. Ruby provides the `#compact` method to remove `nil` values from an array and (since 2.4) has provided a [the same `#compact` method](https://ruby-doc.org/core-2.4.0/Hash.html#method-i-compact) for Hash.
```

> But often I’m looking to remove empty strings or other “blank” objects.

Reviewer: Too many "I's" in this story...

...always pitch it from the perspective of the reader: what's their pain point? How does Active Support help them to solve it?

## Instead of…

…manually looping through the values of a hash to remove `nil`s and empty strings:

```ruby
[1, nil, “3”, “”, []].compact
#=> [1, “3”, “”, []]

[1, nil, “3”, “”, []].reject { |e| e.nil? || e&.empty? }
#=> [1, “3”]

[1, nil, “3”, “”, []].reject(&:blank?)
#=> [1, “3”]

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.compact
#=> {1 => 2, “3” => “4”, 4 => “”, 5 => []}

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.reject { |_, v| v.nil? || v&.empty? }
#=> {1 => 2, “3” => “4”}

{1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.reject { |_, v| v.blank? }
#=> {1 => 2, “3” => “4”}
```

## Use…

…Active Support’s `#compact_blank` method:

```ruby
[1, nil, “3”, “”, []].compact_blank
#=> [1, “3”]

result = {1 => 2, 2 => nil, “3” => “4”, 4 => “”, 5 => []}.compact_blank
#=> {1 => 2, “3” => “4”}
```

The [#compact_blank](https://api.rubyonrails.org/classes/Enumerable.html#method-i-compact_blank) method was [added in Rails 6.1](https://github.com/rails/rails/blob/v6.1.4/activesupport/CHANGELOG.md#rails-610-december-09-2020).

## Why?

> This is one of those “useful Rails things” that feels like it’s part of Ruby once you’re used to it, something that a lot of the Active Support library core extensions have in common.

Reviewer: This sentence is a little messy!

> Even some of the `#reject`-based examples I use [Active Support’s long-standing `#blank?`](https://api.rubyonrails.org/classes/Object.html#method-i-blank-3F) method to make the condition more readable.

Reviewer: This sentence has a typo in it, I think, as I'm not sure what you're trying to say.

## Why not?

Some folks resent the intrusion of the framework into the base concepts of the language.

> However there have definitely been cases where Ruby has adopted a “Active Support-ism”, `Hash#compact` was in Rails, long before it was in the base language.

Reviewer: ```suggestion
However, there have definitely been cases where Ruby has adopted an “Active Support-ism”. `Hash#compact` was in Rails long before it was in the base language.
```
